### PR TITLE
Headless/tahsin/bc 2/bc 207/brands api/testing setup

### PR DESCRIPTION
**Pull request by ghost**
# Description
(Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.)

## Type of change

(Please delete options that are not relevant.)

- [x] Bug fix
- [x] New feature

# How Has This Been Tested?

(Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration)

- [x] Tested by myself
- [ ] Unit test

# Developer Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# Review check list:

- [ ] Separation of Concerns followed (Reviewer name/NaN)
- [ ] Code is in sync with existing code patterns/technologies. (Reviewer name/NaN)
- [ ] Are there any obvious optimizations that will improve performance? (Reviewer name/NaN)
- [ ] Can any logging or debugging code be removed? (Reviewer name/NaN)
- [ ] Are all data inputs checked (for the correct type, length, format, and range) and encoded? (Reviewer name/NaN)
- [ ] Are invalid parameter values handled? (Reviewer name/NaN)
- [ ] Is error handling done the correct way? (Reviewer name/NaN)
- [ ] Are all the edge cases (null, 0, negative) handled properly? (Reviewer name/NaN)

### CODE REVIEW

File: packages/headless/config/database/index.ts [hunk 1]
```diff
@@ -11,16 +11,16 @@ const {
...
-		username: MYSQL_DB_USER! || 'root',
-		password: MYSQL_DB_PASSWORD! || '',
-		database: MYSQL_DB_NAME! || 'bs-commerce-dev',
-		host: MYSQL_DB_HOST! || 'localhost',
-		port: parseInt(MYSQL_DB_PORT!) || 3306,
-		dialect: (MYSQL_DB_DIALECT as Dialect) || 'mysql',
-	},
-	db: DB || 'MONGO',
-};
+    mongodb: {
```
Comment (ismail0946) [inline packages/headless/config/database/index.ts:14]: why are the changes to this file?

File: packages/headless/package.json [hunk 2]
```diff
@@ -1,78 +1,87 @@
...
-		"prettier": "^2.3.2",
-		"source-map-support": "^0.5.20",
-		"supertest": "^6.1.3",
-		"ts-jest": "^27.0.3",
-		"ts-loader": "^9.2.3",
-		"ts-node": "^10.0.0",
-		"tsconfig-paths": "^3.10.1",
-		"typescript": "^4.3.5"
-	}
+  "name": "headless",
```
Comment (ismail0946) [inline packages/headless/package.json:2]: same here

File: packages/headless/src/app.module.ts [hunk 3]
```diff
@@ -6,7 +6,7 @@ import { coreConfig } from 'config/core';
 @Module({
   imports: [
     HelperModule,
-    ...((coreConfig.api === 'GRAPHQL') ? ResolveGraphqlModule() : ResolveRestModule())
+    ...( (coreConfig.api === 'GRAPHQL') ? ResolveGraphqlModule() : ResolveRestModule() )
```
Comment (ismail0946) [inline packages/headless/src/app.module.ts:9]: why are the changes to this file?

File: packages/headless/src/database/database.init.ts [hunk 4]
```diff
@@ -4,19 +4,21 @@ import { connect as connectToMySql } from './mysql/connect';
...
-				await connectToMySql();
-				break;
-			default:
-				throw new Error('No database found to connect');
-		}
-	} catch (err) {
-		console.log(err);
-		console.error('Error connecting to database');
-	}
+    try {
```
Comment (ismail0946) [inline packages/headless/src/database/database.init.ts:7]: why are the changes to this file?

File: packages/headless/src/database/mysql/connect.ts [hunk 5]
```diff
@@ -21,4 +22,5 @@ export const connect = async () => {
             console.log(e.message);
         });
     return sequelize;
+
```
Comment (ismail0946) [inline packages/headless/src/database/mysql/connect.ts:25]: why new line here?

File: packages/headless/src/modules/brands/repositories/index.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import { Brand } from 'src/entity/brand';
+import { BrandDto } from 'src/modules/brands/dto/brandDto';
+import { CreateBrandRequestDto } from 'src/modules/brands/dto/createBrandDto';
+import { GetAllBrandsDto } from 'src/modules/brands/dto/getAllBrandsDto';
+import { UpdateBrandRequestdto, UpdateBrandResponseDto } from 'src/modules/brands/dto/updateBrandDto';
+
+@Injectable()
+export class BrandRepository {
+    constructor(private readonly db: IBrandDatabase) { }
+
```
Comment (ismail0946) [inline packages/headless/src/modules/brands/repositories/index.ts:14]: there are many spaces in the whole PR.

File: packages/headless/src/modules/brands/rest/brand.controller.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  constructor(private brandService: BrandService) {}
+
+  @Get('/')
+  @ApiResponse({
+    status: HttpStatus.OK,
+    description: 'All the brands were fetched successfully',
+    type: GetAllBrandsSuccessResponseDto
+  })
+  @ApiResponse({
+    status: HttpStatus.INTERNAL_SERVER_ERROR,
```
Comment (ismail0946) [inline packages/headless/src/modules/brands/rest/brand.controller.ts:37]: why INTERNAL_SERVER_ERROR?

File: packages/headless/src/modules/brands/services/index.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import { DeleteBrandResponseDto } from "src/modules/brands/dto/deleteBrandDto";
+
+@Injectable()
+
+export class BrandService{
+    constructor( private brandRepo: BrandRepository ){}
+
+    async createBrand( brand: CreateBrandRequestDto ): Promise<CreateBrandResponseDto>{
+            const doesBrandExist = await this.brandRepo.getBrandByName(brand.info.name);
+            if(doesBrandExist) return { error: ErrorMessage.BRAND_ALREADY_EXISTS, errors: null, code: HttpStatus.BAD_REQUEST};
```
Comment (ismail0946) [inline packages/headless/src/modules/brands/services/index.ts:23]: you can write code here without an if-else block.

File: packages/models/src/brand/createBrand.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    code: number;
+    data: CreateBrandRequest
+}
+
+export interface CreateBrandErrorResponse extends ErrorResponse{
+    error: ErrorMessage.CANNOT_CREATE_BRAND | ErrorMessage.BRAND_ALREADY_EXISTS | ErrorMessage.NAME_REQUIRED | ErrorMessage.NAME_BE_VALID;
+}
+
+export const enum ErrorMessage{
+    CANNOT_CREATE_BRAND = 'CANNOT CREATE BRAND',
```
Comment (ismail0946) [inline packages/models/src/brand/createBrand.ts:35]: can be the same Uppercase Error
like 
`CANNOT_CREATE_BRAND = 'CANNOT_CREATE_BRAND'`

File: packages/models/src/brand/deleteBrand.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    code: number;
+    data: Brand
+}
+
+export interface DeleteBrandErrorResponse extends ErrorResponse{
+    error: ErrorMessageDeleteBrand.INVALID_BRAND_ID | ErrorMessageDeleteBrand.CANNOT_DELETE_BRAND;
+}
+
+export const enum ErrorMessageDeleteBrand{
+    INVALID_BRAND_ID = 'NO BRAND WITH SUCH ID',
```
Comment (ismail0946) [inline packages/models/src/brand/deleteBrand.ts:16]: same here

File: packages/models/src/common/errorResponse.ts [hunk 11]
```diff
@@ -5,4 +5,4 @@ export interface ErrorResponse {
 }
 export interface DescriptiveError {
     [key: string]: [string];
-}
\ No newline at end of file
+}
```
Comment (ismail0946) [inline packages/models/src/common/errorResponse.ts:8]: why new line here?

File: packages/headless/config/database/index.ts (hunk 1 above)
Comment (ghost) [reply packages/headless/config/database/index.ts]: dont know how it got changed, maybe pulling from another branch or a test branch of mine did it

File: packages/headless/src/modules/brands/rest/brand.controller.ts (hunk 7 above)
Comment (ghost) [reply packages/headless/src/modules/brands/rest/brand.controller.ts]: get all brands implements no logic or gets no paramter, so if there is any error, what error will it be?
bad request is not an error for a simple get request

File: packages/headless/src/modules/brands/services/index.ts (hunk 8 above)
Comment (ghost) [reply packages/headless/src/modules/brands/services/index.ts]: how? can you please elaborate

File: packages/models/src/brand/createBrand.ts (hunk 9 above)
Comment (ghost) [reply packages/models/src/brand/createBrand.ts]: can you please elaborate?

**Pull request merged**